### DISCUSSION TITLE
Sync randomly fails with CORS error

### DISCUSSION THREAD

**Discussion by jmason888** [standardnotes/forum#3122] [General]
**Describe the bug**
In the web app, after editing notes, sync randomly fails about 50% of the time. When this occurs, sync will eventually succeed after several minutes.  This delay can mean that notes get out of sync with other devices.

When the error occurs, the following is found in the JavaScript console:
```
Access to XMLHttpRequest at 'https://api.standardnotes.com/v1/items' from origin 'https://app.standardnotes.com' has been blocked by CORS policy: Response to preflight request doesn't pass access control check: No 'Access-Control-Allow-Origin' header is present on the requested resource. 
```

Also, when this error occurs, some other account-level actions don't work either: in particular, the Listed menu says "no author found" even though I do have a (private) blog set up.

**To Reproduce**
1. Go to https://app.standardnotes.com/
2. Edit several notes
3. The account icon in the lower left will eventually show red indicating a sync error
4. Open the JavaScript console

**Expected behavior**
Sync should not fail unless the client has lost its internet connection. My internet connection is working at the time (other pages work), and what's more, https://app.standardnotes.com/ works too. It's just sync that randomly fails.

**Desktop (using the Web app):**
 - OS: ChromeOS
 - Browser: Chrome
 - Version: 109.0.5414.21

**Comment by effieeee** [comment]
Hi @jmason888, do you encounter similar syncing issues using other browsers, or with other app platforms like the desktop app or mobile app?

Please also send us a message at help@standardnotes.com, and mention your account email 🙏

**Comment by jmason888** [comment]
I do most of the heavy editing on the desktop, and my desktop use is almost exclusively on a Chromebook.

I use the Android app, and I only once had it give sync failures... and that was on a day when there were other connectivity problems with Standard Notes (like errors from CloudFlare), so I imagine that it was an unrelated issue.

But in the browser, I see the issue pretty regularly, on various different ISPs (it's a laptop, so it gets around...)
If I only do a handful of edits, I'm OK, but once I start to do any serious editing (a lot of changes) it starts to give these sync errors (with CORS errors showing up in the JS console)... and then I have to wait a few minutes after my last edit for the sync to finally succeed.  It's almost as if there was some rate-limiting or load-balancing happening the back-end connection with the API, and some of the servers/load-balancers aren't sending the right CORS headers while others are...

**Comment by jmason888** [comment]
I took a look in more detail, leaving the network waterfall display open in the Developer Tools while the app is open. What happens is this:

After logging in, the app refreshes the page and clears the waterfall display, then I see:
- The app requests a whole bunch of resources, which work fine
- After those have settled, the app requests https://api.standardnotes.com/v1/items repeatedly in the background
- These requests are pretty rapid: the waterfall logged about 1300 of them in the first 10 minutes I had the app open
- After a few (1 to 5) minutes, the "items" requests start failing with CORS errors.
- The app backs off and requests them less often
- After another 4 or 5 minutes, the "items' requests start succeeding again, and restart their rapid pace

It doesn't actually seem to depend on how much editing I am doing; it just seems to come and go as long as I have the app open.

**Comment by jmason888** [comment]
@effieeee I sent that email the other day.

Now I'm wondering if this is CloudFlare blocking some requests. Developer Tools records the following when the request to https://api.standardnotes.com/v1/items fails with the CORS errors:

> **General:**
> Request URL: https://api.standardnotes.com/v1/items
> Request Method: OPTIONS
> Status Code: 403 
> Remote Address: 34.225.224.245:443
> Referrer Policy: no-referrer-when-downgrade
> 
> **Response Headers:**
> content-length: 520
> content-type: text/html
> date: Mon, 19 Dec 2022 22:53:02 GMT
> server: awselb/2.0

Note the _content-type: text/html_: this is a call to an API endpoint; it would normally be returning _application/json_ or something like that.

On the other hand, if CloudFlare were intercepting the request and blocking it (for whatever reason), they wouldn't know it's an API page and return an HTML page for the user to see. And 520 bytes is about the right length for a quick HTML "403: go away--request blocked" page.

**Comment by effieeee** [comment]
Hi @jmason888, apologies for the late response, and thank you for providing those added details! One of the devs had an initial look at the network request log you gave as well, though I'm afraid we're still rather unsure of what may be causing the random sync issue you're experiencing. I'll try to keep you posted with some updates here, or they might also reach out directly on this thread 🙏
